### PR TITLE
React: Switch react-docgen-typescript-loader to react-docgen-typescript-plugin

### PR DESCRIPTION
**Pull request by hipstersmoothie**
## What I did

Typescript docgen for react was super slow!
This pull request switches from `react-docgen-typescript-loader` to `react-docgen-typescript-plugin` 

*Differences:*

- In a webpack plugin instead of loader
- This allows for the typescript docgen to happen all at once and be non-blocking. (as a loader it synchronously executes for each file)
- It also allows the plugin to only create the typescript compiler and docgen parser once instead of for every file

In my own project this resulted in a huge speed boost:

**Before:**

Screen Shot 2020-06-09 at 12 14 43 PM

**After:**

Screen Shot 2020-06-09 at 1 59 57 PM

## How to test

- Is this testable with Jest or Chromatic screenshots? no
- Does this need a new example in the kitchen sink apps? no
- Does this need an update to the documentation? yes

If your answer is yes to any of these, please make sure to include it in your PR.

### CODE REVIEW

**Review by shilman — CHANGES_REQUESTED** [review]
This looks amazing, I can't wait to kick the tires.

This is is a pretty big change so expect some back and forth before we get it merged. First question: what about these: 

https://github.com/storybookjs/storybook/blob/next/lib/core/src/server/config/defaults.js#L4-L7

**Comment by shilman** [comment]
cc @mrmckeb

**Comment by hipstersmoothie** [comment]
> what about these:

Both of those options will be respected.

https://www.npmjs.com/package/react-docgen-typescript-plugin

The only thing I changed with the options for the plugin are `skipPropsWithName` `skipPropsWithoutDoc`, and `compilerOptions`. 

## `skipPropsWithName` `skipPropsWithoutDoc`

Instead of creating and extra options we just pass [`propFilter` to `react-docgen-typescript`](https://github.com/styleguidist/react-docgen-typescript#parseroptions)

https://github.com/strothj/react-docgen-typescript-loader 

## `compilerOptions`

I can add this in if everyone wants it, but it seems way simpler to me to just point at a tsconfig in the repo.

File: addons/docs/src/lib/sbtypes/proptypes/convert.ts [hunk 1]
```diff
@@ -43,8 +43,7 @@ export const convert = (type: PTType): SBType | any => {
     case 'elementType':
     default: {
       if (name?.indexOf('|') > 0) {
-        // react-docgen-typescript-loader doesn't always produce proper
-        // enum types, possibly due to https://github.com/strothj/react-docgen-typescript-loader/issues/81
+        // react-docgen-typescript-plugin doesn't always produce proper
```
Comment (mrmckeb) [inline addons/docs/src/lib/sbtypes/proptypes/convert.ts:46]: This comment seems to have only been partially removed?

File: app/react/src/server/framework-preset-react-docgen.ts [hunk 2]
```diff
@@ -27,24 +30,13 @@ export function babel(config: TransformOptions, { typescriptOptions }: Storybook
 
 export function webpackFinal(config: Configuration, { typescriptOptions }: StorybookOptions) {
   const { reactDocgen, reactDocgenTypescriptOptions } = typescriptOptions;
-  if (reactDocgen !== 'react-docgen-typescript') return config;
+
+  if (reactDocgen !== 'react-docgen-typescript') {
```
Comment (mrmckeb) [inline app/react/src/server/framework-preset-react-docgen.ts:34]: Why was this changed?

**Review by mrmckeb — CHANGES_REQUESTED** [review]
A few questions @hipstersmoothie, but looks good overall.

> I can add this in if everyone wants it, but it seems way simpler to me to just point at a tsconfig in the repo.
You mean you would look for a root config?

**Comment by mrmckeb** [comment]
@shilman when this merges, we can make this change in CRA too.

**Comment by shilman** [comment]
The chromatic snapshots look pretty good. ❤️  @hipstersmoothie please take a look at the "enum"-related snapshot diffs. My enum handling code for RDTL is a hack, and it got "slightly worse" in your version. I'm not concerned about the regressions since it didn't work well in the first place. But I wanted to point it out because if you're going to do more hacking on `react-docgen-typescript`/(`-plugin`), that would be a really great area to give some attention to.

File: app/react/src/server/framework-preset-react-docgen.ts (hunk 2 above)
Comment (ndelangen) [reply app/react/src/server/framework-preset-react-docgen.ts:34]: I personally prefer the *new* code

**Review by ndelangen — APPROVED** [review]

**Comment by shilman** [comment]
@hipstersmoothie that was fast! and awesome!! I'll try to kick the tires on this today and hopefully merge.

File: addons/docs/src/lib/sbtypes/proptypes/convert.ts (hunk 1 above)
Comment (mrmckeb) [reply addons/docs/src/lib/sbtypes/proptypes/convert.ts:46]: I understand, but one line was removed and now it doesn't make sense, right?
> // react-docgen-typescript-plugin doesn't always produce proper
> // this hack tries to parse out values from the string and should be
> // removed when RDTL gets a little smarter about this

File: app/react/src/server/framework-preset-react-docgen.ts (hunk 2 above)
Comment (mrmckeb) [reply app/react/src/server/framework-preset-react-docgen.ts:34]: I was more pointing out that this change isn't related to any of the work. I think it's good practice to try to only change things related to the task at hand, unless something is being improved. But if you think the change is meaningful, then I'm OK with that.

**Comment by shilman** [comment]
Looking fantastic. Here are the numbers I'm seeing for `official-storybook` on my machine:

* react-docgen: 17 s for manager and 21 s for preview
* react-docgen-typescript (before): 36 s for manager and 46 s for preview
* react-docgen-typescript (after): 15 s for manager and 31 s for preview

**Comment by mrmckeb** [comment]
@hipstersmoothie It looks like this has introduced a regression, as `react-docgen-typescript-plugin` loads for all files (not just TypeScript), which means it is also injected for projects that aren't using TypeScript at all - and fails, as the `typescript` package is not found.

You can see here that we used to only target `tsx?` files. Is it possible to resolve this within the plugin?

**Comment by hipstersmoothie** [comment]
I actually just released an update with this functionality. 0.3.0 should only include tsx files.

**Comment by mrmckeb** [comment]
That's great news, and I'm guessing it only loads TypeScript as a dependency when `ts` and `tsx` files are found?

**Comment by hipstersmoothie** [comment]
Are you asking if I can conditionally require typescript if it’s installed? I’m assuming the loader only ever got loaded if a typescript file was found

**Comment by mrmckeb** [comment]
Yes, that was the previous behaviour. With this change, the plugin is always invoked - but we know that TypeScript may not be installed at all times. Do you conditionally require TypeScript now?

If not, we may need to revert this temporarily, as it will break in any projects that don't have a `typescript` dependency.

**Comment by hipstersmoothie** [comment]
PR incoming https://github.com/hipstersmoothie/react-docgen-typescript-plugin/releases/tag/v0.4.0
